lib/ramble/ramble/pipeline.py

```diff
@@ -0,0 +1,338 @@
+                            'workspace first.\n' + \
+                            'Then ensure its spack configuration is ' + \
+                            'properly configured.'
+            with self.workspace.logger.force_echo():
+                tty.die(error_message)
+
+    def _prepare(self):
+        """Perform preparation for pipeline execution"""
+        pass
+
+    def _execute(self):
+        """Hook for executing the pipeline"""
+        for exp, app_inst, count in self._experiment_set.filtered_experiments(self.filters):
+            exp_log_path = app_inst.experiment_log_file(self.log_dir)
+            exp_log_stream = open(exp_log_path, 'a+')
+
+            tty.msg(f'    Experiment {count}:',
+                    f'        name: {exp}',
+                    f'        log file: {exp_log_path}')
+
+            for phase in app_inst.get_pipeline_phases(self.name, self.filters.phases):
+                with app_inst.logger(exp_log_stream):
+                    app_inst.run_phase(phase, self.workspace)
+
+    def _complete(self):
+        """Hook for performing pipeline actions after execution is complete"""
+        pass
+
+    def run(self):
+        """Run the full pipeline"""
+        self.log_stream = open(self.log_path, 'a+')
+
+        tty.msg('Streaming details to log:')
+        tty.msg(f'  {self.log_path}')
+        if self.workspace.dry_run:
+            cprint('@*g{      -- DRY-RUN -- DRY-RUN -- DRY-RUN -- DRY-RUN -- DRY-RUN --}')
+
+        experiment_count = self._experiment_set.num_filtered_experiments(self.filters)
+        experiment_total = self._experiment_set.num_experiments()
+        tty.msg(f'  {self.action_string} {experiment_count} out of '
+                f'{experiment_total} experiments:')
+
+        with self.workspace.logger(self.log_stream, echo=self.workspace.dry_run):
+            self._validate()
+            self._prepare()
+
+        self._execute()
+
+        with self.workspace.logger(self.log_stream, echo=self.workspace.dry_run):
+            self._complete()
+
+        self.log_stream.close()
+
+
+class AnalyzePipeline(Pipeline):
+    """Class for the analyze pipeline"""
+
+    name = 'analyze'
+
+    def __init__(self, workspace, filters, output_formats=['text'], upload=False):
+        workspace_success = {
+            namespace.success: ramble.config.config.get_config(namespace.success)
+        }
+
+        workspace.extract_success_criteria('workspace', workspace_success)
+
+        super().__init__(workspace, filters)
+        self.action_string = 'Analyzing'
+        self.output_formats = output_formats
+        self.require_inventory = True
+        self.upload_results = upload
+
+    def _prepare(self):
+        super()._construct_hash()
+        super()._prepare()
+
+    def _complete(self):
+        self.workspace.dump_results(output_formats=self.output_formats)
+
+        # FIXME: this will fire the analyze logic of twice currently
```

Comment:
This is in the wrong place, I think (or at least the upload_results doesn't trigger analyze) 

We can probably just remove it..